### PR TITLE
Tests of Query API cursor-based pagination

### PR DESCRIPTION
**Pull request by garethsb-sony**
We're attempting to convert an internal unit test suite into API-based tests for the 'paged' trait.

This is the beginning:

1. Check all required response headers are present
2. Check the most recently POSTed resource is at the top of the initial page of results
3. Check paging.limit restricts the response to the specified number of results

More to come...

### CODE REVIEW

**Comment by garethsb-sony** [comment]
We think this is now ready for review and testing against other implementations.

The biggest problem I have with the current approach is this:
https://github.com/AMWA-TV/nmos-testing/blob/d56ca4a4307e4ebb560a02245bf11d8829c9769b/IS0402Test.py#L561-L564

Thoughts on that would also be appreciated!

**Comment by garethsb-sony** [comment]
> ```
> # Another alternative would be to use local timestamps [...] with approximate matching? 
> ```

There's a proof-of-concept at https://github.com/garethsb-sony/nmos-testing/commit/1b8d58a1748c09ea6472681b618eef2e853b712d

But it's very susceptible to Test Suite and Registry clocks not being synchronised, so I haven't pushed it to this PR branch.

File: IS0402Test.py [hunk 1]
```diff
@@ -507,31 +412,595 @@ def test_20(self):
...
+        before = self.is04_query_utils.get_TAI_time()
+        after = self.is04_query_utils.get_TAI_time(1)
+
+        # Specifying since after until is a bad request
+        valid, response, query_string = self.do_paged_request(since = after, until = before)
+
+        if not valid:
+            raise NMOSTestException(test.FAIL("Query API did not respond as expected, "
+                                              "for query: {}".format(query_string)))
+        elif response.status_code != 400:
```
Comment (andrewbonney) [inline IS0402Test.py]: Does this one also need a 501 check?

File: IS0402Test.py [hunk 2]
```diff
@@ -507,31 +412,595 @@ def test_20(self):
...
+            if not valid:
+                raise NMOSTestException(test.FAIL("Cannot GET the POSTed sample data. Cannot execute test: "
+                                                  "{}".format(r)))
+            elif r.status_code != 200:
+                raise NMOSTestException(test.FAIL("Cannot GET the POSTed sample data. Cannot execute test: "
+                                                  "{} {}".format(r.status_code, r.text)))
+
+            try:
+                if r.json()[0]["id"] != node_data["id"]:
+                    raise NMOSTestException(test.FAIL("Query API response did not have the most recently POSTed node first"))
```
Comment (andrewbonney) [inline IS0402Test.py]: Is it worth checking here if any paging headers are returned first? You won't get a 501 before this point, but if a registry doesn't implement pagination there isn't necessarily an ordering guarantee.

File: IS0402Test.py (hunk 1 above)
Comment (garethsb-sony) [reply IS0402Test.py]: Thanks. Yes. I'll hack our Registry to return 501 so I'll pick this and any similar misses up.

File: IS0402Test.py (hunk 2 above)
Comment (garethsb-sony) [reply IS0402Test.py]: > "if a registry doesn't implement pagination there isn't necessarily an ordering guarantee"

😲 Oh my, that hadn't even occurred to me!

File: IS0402Test.py (hunk 1 above)
Comment (garethsb-sony) [reply IS0402Test.py]: Resolved by 3837909c27935738d5afe3db8d9be1a73f485394

File: IS0402Test.py (hunk 2 above)
Comment (garethsb-sony) [reply IS0402Test.py]: Resolved by 3837909c27935738d5afe3db8d9be1a73f485394

**Comment by garethsb-sony** [comment]
I've now tested this against a Registry hacked to remove support for pagination and/or basic queries, and the Test Suite produces the expected list of "Not Implemented" results.

**Comment by andrewbonney** [comment]
One further thing that may need resolving. I'd got slightly confused as many tests were correctly coming back as not implemented for our registry, but one wasn't. We don't currently support the 501 code for pagination as it wasn't explicitly in the spec until v1.3. I don't see a major issue adding/backporting this in our implementation, but as it came up I thought it might be worth noting that others could hit this:

In test_26, I receive 'Query API responded with wrong HTTP code, for query: ?paging.since=1550678485:483987092&paging.until=1550678484:483979463' rather than 'Query API response did not include any pagination headers. Query APIs should support pagination for scalability.' which is returned by the others.

**Comment by garethsb-sony** [comment]
@andrewbonney Did you mean test_26? And did you test after the last two commits, because I resolved something that sounds like what you saw?

**Comment by andrewbonney** [comment]
Sorry, I meant 21_6. Yes, I'm still seeing the issue after those changes.

**Comment by garethsb-sony** [comment]
So, what code is your v1.2 Query API reporting? If it's 200, isn't that wrong because it means it was ignoring the paging parameters? It should have returned at least a 400 to indicate it doesn't understand the request?

**Comment by garethsb-sony** [comment]
Hmm, OK, I see; QueryAPI.raml at v1.2.x doesn't offer anything except 200 for either the 'paged' trait or explicitly for each of the GET /nodes, /devices/, etc. Yuck.

So yes, I agree, the "no pagination headers" warning would be better...

**Comment by andrewbonney** [comment]
Yeah, it essentially gets treated as a basic query parameter at present. The 501 is the solution, just a pity we didn't include it at the time.

**Comment by garethsb-sony** [comment]
OK. I do remember discussing whether "paging." and "query." ought to be treated as namespaces within which unrecognised parameter names ought to result in 400/501, rather than being treated as basic query parameters.

**Comment by garethsb-sony** [comment]
OK, a3a8a3c023102850f5733a812dc2206c3b91e956 should resolve that last comment, @andrewbonney.
I've now tested with a Registry that supports pagination, one that declares that it doesn't with a 501, and one that just ignores 'paging' parameters. I think these tests are as good as I can make them for now!

**Comment by garethsb-sony** [comment]
Thanks and sorry, Andrew, I just made one more commit. Please would you mind reviewing the one new test for correct encoding of query parameter values in Link response headers?

**Pull request merged**